docs/best-practices.md

```diff
@@ -0,0 +1,799 @@
+        Promise::new(receiver_id.clone())
+            .function_call(
+                b"mult",
+                json!({ "a": a, "b": b }).to_string().as_bytes(),
+                deposit,
+                gas,
+            )
+    }
+
+    pub fn sum(a: U128, b: U128, receiver_id: &AccountId, deposit: Balance, gas: Gas) -> Promise {
+        // ...
+    }
+}
+```
+
+Let's assume the calculator is deployed on `calc.near`, we can use the following:
+
+```rust
+const CALCULATOR_ACCOUNT_ID: &str = "calc.near";
+const NO_DEPOSIT: Balance = 0;
+const BASE_GAS: Gas = 5_000_000_000_000;
+
+#[near_bindgen]
+impl Contract {
+    pub fn sum_a_b(&mut self, a: U128, b: U128) -> Promise {
+        let calculator_account_id: AccountId = CALCULATOR_ACCOUNT_ID.to_string();
+        ext_calculator::sum(a, b, &calculator_account_id, NO_DEPOSIT, BASE_GAS)
+    }
+}
+```
+
+## Reuse crates from `near-sdk`
+
+`near-sdk` re-exports the following crates:
+
+- `borsh`
+- `base64`
+- `bs58`
+- `serde`
+- `serde_json`
+
+Most common crates include `borsh` which is needed for internal STATE serialization and
+`serde` for external JSON serialization.
+
+When marking structs with `serde::Serialize` you need to use `#[serde(crate = "near_sdk::serde")]`
+to point serde to the correct base crate.
+
+```rust
+/// Import `borsh` from `near_sdk` crate 
+use near_sdk::borsh::{self, BorshDeserialize, BorshSerialize};
+/// Import `serde` from `near_sdk` crate 
+use near_sdk::serde::{Serialize, Deserialize};
+
+/// Main contract structure serialized with Borsh
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize, PanicOnDefault)]
+pub struct Contract {
+    pub pair: Pair,
+}
+
+/// Implements both `serde` and `borsh` serialization.
+/// `serde` is typically useful when returning a struct in JSON format for a frontend.
+#[derive(Serialize, Deserialize, BorshDeserialize, BorshSerialize)]
+#[serde(crate = "near_sdk::serde")]
+pub struct Pair {
+    pub a: u32,
+    pub b: u32,
+}
+
+#[near_bindgen]
+impl Contract {
+    #[init]
+    pub fn new(pair: Pair) -> Self {
+        Self {
+            pair,
+        }
+    }
+
+    pub fn get_pair(self) -> Pair {
+        self.pair
+    }
+}
+```
+
+## `std::panic!` vs `env::panic`
+
+- `std::panic!` panics the current thread. It uses `format!` internally, so it can take arguments.
+  SDK sets up a panic hook, which converts the generated `PanicInfo` from `panic!` into a string and uses `env::panic` internally to report it to Runtime.
+  This may provide extra debugging information such as the line number of the source code where the panic happened.
+
+- `env::panic` directly calls the host method to panic the contract.
+  It doesn't provide any other extra debugging information except for the passed message.
+
+## In-memory `HashMap` vs persistent `UnorderedMap`
+
+- `HashMap` keeps all data in memory. To access it, the contract needs to deserialize the whole map.
```

Comment:
maybe it's fine to use this as a comparison, although I think this section could be moved also, but probably should compare with `BTreeMap`, as it's much better in most cases in wasm contexts (BTreeMaps are slow with a large number of elements, but that's why you have the collections anyway)

Reply:
I'm afraid I can't write a good comparison on this one, but have moved a large chunk of this section into the Collections page https://github.com/near/sdk-docs/pull/37/commits/cb90e333d8eafabf6c3debaeb83e1e9f7a6420c5